pallets/phala/src/compute/wrapped_balances.rs

```diff
@@ -0,0 +1,505 @@
+			aye_amount: BalanceOf<T>,
+			nay_amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// user's `FinanceAccount` does not exist in storage: `StakerAccounts`
+		StakerAccountNotFound,
+		/// Trying to unwrap more than the available balance
+		RedeemAmountExceedsAvaliableStake,
+		/// Trying to vote more than the available balance
+		VoteAmountLargerThanTotalStakes,
+		/// The vote is not currently on going
+		ReferendumInvalid,
+		/// The vote is now on going and the W-PHA used in voting can not be unlocked
+		ReferendumOngoing,
+		/// The Iteration exceed the max limitaion
+		IterationsIsNotVaild,
+	}
+
+
+
+	impl<T: Config> rmrk_traits::TransferHooks<T::AccountId, u32, u32> for Pallet<T>
+	where
+		BalanceOf<T>: sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+		T: pallet_uniques::Config<CollectionId = CollectionId, ItemId = NftId>,
+		T: pallet_assets::Config<AssetId = u32, Balance = BalanceOf<T>>,
+		T: pallet_democracy::Config<Currency = <T as crate::PhalaConfig>::Currency>,
+		T: Config + vault::Config,
+	{
+		fn pre_check(_sender: &T::AccountId, _collection_id: &CollectionId, _nft_id: &NftId) -> bool 
+		{
+			match base_pool::pallet::PoolCollections::<T>::get(_collection_id) {
+				Some(pid) => {
+					if let Some(net_value) = Pallet::<T>::get_net_value((*_sender).clone()).ok() {
+						let property_guard = base_pool::Pallet::<T>::get_nft_attr_guard(*_collection_id, *_nft_id)
+							.expect("get nft should not fail: qed.");
+						let property = &property_guard.attr;
+						let account_status = match StakerAccounts::<T>::get(_sender) {
+							Some(account_status) => account_status,
+							// Already get_net_value, in fact never get here.
+							None => return true,
+						};			
+						let pool_proxy = base_pool::Pallet::<T>::pool_collection(pid)
+								.expect("get pool should not fail: qed.");
+						let basepool = &match pool_proxy {
+							PoolProxy::Vault(p) => p.basepool,
+							PoolProxy::StakePool(p) => p.basepool,
+						};
+						if let Some(price) = basepool.share_price() {
+							let nft_value = bmul(property.shares, &price);
+							if account_status.locked + nft_value > net_value {
+								return false;
+							}
+						} else {
+							return true;
+						}		
+					}
+				},
+				None => return true,
+			};
+
+			true
+		}
+		fn post_transfer(
+			sender: &T::AccountId,
+			recipient: &T::AccountId,
+			collection_id: &CollectionId,
+			nft_id: &NftId,
+		) -> bool {
+			match base_pool::pallet::PoolCollections::<T>::get(collection_id) {
+				Some(pid) => {
+					let pool_proxy = base_pool::Pallet::<T>::pool_collection(pid)
+						.expect("already checked exist; qed.");
+					let pool_type = match pool_proxy {
+						PoolProxy::Vault(res) => PoolType::Vault,
+						PoolProxy::StakePool(res) => PoolType::StakePool,
+					};
+					base_pool::Pallet::<T>::merge_or_init_nft_for_staker(*collection_id, recipient.clone(), pid, pool_type);
```

Comment:
If the nft is all transferred out, will it create a new NFT with 0 share?

Reply:
yes it will be

Reply:
Isn't it better to avoid creating a new zero NFT?

Reply:
In initial design, the function is also used to init a new zero nft,and it always returns result(nft_id) if nothing goes wrong, i think it maybe too late to change that lot.

Reply:
Maybe add a new function called "maybe_merge_nft" with the almost the same code as `merge_or_init_nft_for_staker` but just don't create zero NFT in the end?